### PR TITLE
Cli commands

### PR DESCRIPTION
**Pull request by gigabo**
Now that the CLI can be [installed globally](https://github.com/redfin/react-server/pull/484) it can be used to bootstrap a React Server site.

This is totally new functionality that's only tangentially related to the previous "start server" behavior.  So... I've added "commands".

```
$ react-server
Usage: react-server <command> [options]

Commands:
  start     Start the server
  compile   Compile static assets
  init      Initialize a React Server site
  add-page  Add a page to an existing site

Options:
  --routes-file        The routes file to load. Default is 'routes.js'.
  -p, --port           Port to start listening for react-server. Default is 3000.
...
```

The new quick start in the CLI doc looks like:

```bash
$ npm install -g react-server-cli
$ react-server init
$ react-server add-page '/' Homepage
$ react-server start
```

This is a pretty big change to the interface of `react-server-cli`.

@aickin - How does this look to you?

### CODE REVIEW

File: docs/testing-ports.md [hunk 1]
```diff
@@ -4,7 +4,7 @@ simultaneously, so port conflicts will fail the tests.  Make sure
 to update this manifest when adding a test that starts a server.
 
 ## generator-react-server 
-3010, 3011
+3000, 3001
```
Comment (doug-wade) [inline docs/testing-ports.md]: doesn't 3001 conflict with react-server-integration-test?

Comment (gigabo) [inline docs/testing-ports.md]: I don't think so.  I think `react-server-integration-test` uses a dummy static asset server.

File: docs/guides/react-server-cli.md [hunk 2]
```diff
@@ -1,113 +1,198 @@
...
-### Ways to add options
+### Routes
+
+This is where URLs are mapped to pages.  It's also where global middleware
+(applied to all pages in the site) is defined.
+
+By default this is created at the site's root directory as `routes.json`.
 
-There are three ways to pass options to the CLI, through the command line, `.reactserverrc` JSON files, or as a `reactServer` entry in `package.json` files. It searches for options this way:
+```json
```
Comment (doug-wade) [inline docs/guides/react-server-cli.md]: I feel like the routes.json format should be stored in the docs, and this should be a link to those docs instead of a code snippet

File: docs/guides/react-server-cli.md [hunk 3]
```diff
@@ -1,113 +1,198 @@
...
+Smart defaults are the goal, and `react-server-cli` has two base modes:
+**development** and **production**. `react-server-cli` will determine which
+base mode it's in by looking at the NODE_ENV environment variable. If it's not
+"production", then `react-server-cli` will assume we are in development mode.
+
+### Ways to add options
+
+There are three ways to pass options to the CLI, through the command line,
+`.reactserverrc` JSON files, or as a `reactServer` entry in `package.json`
+files. If there's no config file (or package.json config) in the current
```
Comment (doug-wade) [inline docs/guides/react-server-cli.md]: which takes precendence, `.reactserverrc` or `package.json`?

File: docs/testing-ports.md (hunk 1 above)
Comment (gigabo) [inline docs/testing-ports.md]: Nope, you're right. 😞 
image

File: docs/guides/react-server-cli.md (hunk 3 above)
Comment (gigabo) [inline docs/guides/react-server-cli.md]: I don't actually know.  I had totally forgotten about the `package.json` configuration until I started editing this doc. 😁

File: docs/guides/react-server-cli.md [hunk 4]
```diff
@@ -132,58 +219,72 @@ Minify client JavaScript and CSS.
...
 Defaults to **null**.
 
 #### --https
 
-If true, the server will start up using https with a self-signed certificate. Note that browsers do not trust self-signed certificates by default, so you will have to click through some warning screens. This is a quick and dirty way to test HTTPS, but it has some limitations and should never be used in production. Requires OpenSSL to be installed.
+If true, the server will start up using https with a self-signed certificate.
+Note that browsers do not trust self-signed certificates by default, so you
+will have to click through some warning screens. This is a quick and dirty way
+to test HTTPS, but it has some limitations and should never be used in
+production. Requires OpenSSL to be installed.
```
Comment (doug-wade) [inline docs/guides/react-server-cli.md]: `brew install openssl`?

File: docs/guides/react-server-cli.md (hunk 2 above)
Comment (gigabo) [inline docs/guides/react-server-cli.md]: Huh?  Isn't this the docs?

File: docs/guides/react-server-cli.md [hunk 5]
```diff
@@ -192,20 +293,30 @@ Shows command line options.
...
-		jsPort: 3001,
-		hot: true,
-		minify: false,
-		compileOnly: false,
-		jsUrl: null,
-		longTermCaching: true,
-})
+import {run} from "react-server-cli"
+
+start({
```
Comment (doug-wade) [inline docs/guides/react-server-cli.md]: run?

File: docs/guides/react-server-cli.md [hunk 6]
```diff
@@ -192,20 +293,30 @@ Shows command line options.
...
 ```
 
-All of the values in the second argument are optional, and they have the same defaults as the corresponding CLI arguments explained above. (Also note that if an option isn't present, the programmatic API will search for a config file in the same way as the CLI.)
+The module also exports a `parseCliArgs` function that will let you implement
+your own CLI:
+
+```javascript
+import {run, parseCliArgs} from "react-server-cli"
+
+start(parseCliArgs());
```
Comment (doug-wade) [inline docs/guides/react-server-cli.md]: run?

File: docs/guides/react-server-cli.md (hunk 4 above)
Comment (gigabo) [inline docs/guides/react-server-cli.md]: Oh... I did a bunch of long-line shortening here, so I could see what I was editing.  I should have done that in a dedicated commit.

File: packages/react-server-cli/src/ConfigurationError.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+// Can't use `instanceof` with babel-ified subclasses of builtins.
+//
+//   https://phabricator.babeljs.io/T3083
+//
+// Gotta do this the old-fashioned way. :p
+//
+export default function ConfigurationError(message) {
```
Comment (doug-wade) [inline packages/react-server-cli/src/ConfigurationError.js]: We might want to consider using [VError](https://github.com/davepacheco/node-verror) for our custom error classes

File: packages/react-server-cli/src/commands/compile.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import compileClient from "../compileClient"
+import handleCompilationErrors from "../handleCompilationErrors";
+import setupLogging from "../setupLogging";
+import logProductionWarnings from "../logProductionWarnings";
+
+const logger = require("react-server").logging.getLogger(__LOGGER__);
```
Comment (doug-wade) [inline packages/react-server-cli/src/commands/compile.js]: I thought we were trying to set the pattern of `import {logging} from 'react-server'; const logger = loggin.getLogger(__LOGGER__);`?  Or is that a non-goal?

File: packages/react-server-cli/src/commands/init.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
+import _ from "lodash";
+import fs from "fs";
+import {spawnSync} from "child_process";
```
Comment (doug-wade) [inline packages/react-server-cli/src/commands/init.js]: I'd prefer spawn, personally

File: packages/react-server-cli/src/commands/init.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import chalk from "chalk";
+import fileExists from "../fileExists";
+import ConfigurationError from "../ConfigurationError";
+
+const DEPENDENCIES = [
+	"react-server",
+	"babel-preset-react-server",
+
+	// TODO: Modernize our peer deps and remove versions here.
```
Comment (doug-wade) [inline packages/react-server-cli/src/commands/init.js]: me too thanks

File: packages/react-server-cli/src/commands/init.js (hunk 9 above)
Comment (gigabo) [inline packages/react-server-cli/src/commands/init.js]: There's no concurrency among subprocesses here.  Easier flow of control this way.

File: packages/react-server-cli/src/commands/init.js [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+const DEV_DEPENDENCIES = [
+
+	// TODO: These, too.
+	"webpack-dev-server@~1.13.0",
+	"webpack@^1.13.1",
+]
+
+const CONFIG = {
+	"routes.json": {
+		middleware: [],
```
Comment (doug-wade) [inline packages/react-server-cli/src/commands/init.js]: should we init with a single page/route?

Comment (gigabo) [inline packages/react-server-cli/src/commands/init.js]: Maybe?  My thought was if you want a page you can generate it with `react-server add-page`.  If you don't then this way you won't have to go in and delete something you didn't ask for.

File: packages/react-server-cli/src/commands/init.js [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+	},
+	".babelrc": {
+		presets: ["react-server"],
+	},
+}
+
+export default function init(){
+
+	if (!fileExists("package.json")) {
+		throw new ConfigurationError("Missing package.json.  Please run `npm init` first.");
```
Comment (doug-wade) [inline packages/react-server-cli/src/commands/init.js]: `spawnSync('npm init --yes')` seems more reasonable than a config error

Comment (gigabo) [inline packages/react-server-cli/src/commands/init.js]: Hmm... yeah, I like it.

File: packages/react-server-cli/src/commands/compile.js (hunk 8 above)
Comment (gigabo) [inline packages/react-server-cli/src/commands/compile.js]: Nice catch.

File: packages/react-server-cli/src/fileExists.js [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import fs from "fs";
+
+export default function fileExists(fn) {
```
Comment (doug-wade) [inline packages/react-server-cli/src/fileExists.js]: https://nodejs.org/api/fs.html#fs_fs_accesssync_path_mode ?

File: packages/react-server-cli/src/handleCompilationErrors.js [hunk 14]
```diff
@@ -0,0 +1,28 @@
+import reactServer from "./react-server";
+
+const logger = reactServer.logging.getLogger(__LOGGER__);
```
Comment (doug-wade) [inline packages/react-server-cli/src/handleCompilationErrors.js]: that thing with the destructured import thingy

File: packages/react-server-cli/src/handleCompilationErrors.js [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+
+// takes in the err and stats object returned by a webpack compilation and returns
+// an error object if something serious happened, or null if things are ok.
+export default function handleCompilationErrors (err, stats){
+	if (err) {
+		logger.error("Error during webpack build.");
+		logger.error(err);
+		return new Error(err);
```
Comment (doug-wade) [inline packages/react-server-cli/src/handleCompilationErrors.js]: Maybe worth adding a `CompilationError`?

File: packages/react-server-cli/src/fileExists.js (hunk 13 above)
Comment (gigabo) [inline packages/react-server-cli/src/fileExists.js]: ^ That still throws.  This is just a helper that returns boolean.

File: packages/react-server-cli/src/logProductionWarnings.js [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+		}
+
+		if (!minify) {
+			logger.warning("-- Minification is disabled. To enable, set minify to true (--minify at the command-line) or set NODE_ENV=production.");
+		}
+
+		if (!longTermCaching) {
+			logger.warning("-- Long-term caching is disabled. To enable, set longTermCaching to true (--long-term-caching at the command-line) or set NODE_ENV=production to turn on.");
+		}
+		logger.info(`NODE_ENV is set to ${process.env.NODE_ENV}`); //eslint-disable-line no-process-env
```
Comment (doug-wade) [inline packages/react-server-cli/src/logProductionWarnings.js]: `${process.env.NODE_ENV ? process.env.NODE_ENV : 'development'}`?

File: packages/react-server-cli/src/handleCompilationErrors.js (hunk 15 above)
Comment (gigabo) [inline packages/react-server-cli/src/handleCompilationErrors.js]: Meh.  The only reason for the `ConfigurationError` is to provide a clean UI when we short circuit out due to a simple problem (file already exists, etc).  For ordinary errors where stack trace is actually desirable I don't see any need to subclass.

File: packages/react-server-cli/src/parseCliArgs.js [hunk 17]
```diff
@@ -119,6 +136,14 @@ const sslize = argv => {
 		}
 	}
 
+	if (argv.https && (argv.httpsKey || argv.httpsCert || argv.httpsCa || argv.httpsPfx || argv.httpsPassphrase)) {
+		throw new Error("If you set https to true, you must not set https-key, https-cert, https-ca, https-pfx, or https-passphrase.");
```
Comment (doug-wade) [inline packages/react-server-cli/src/parseCliArgs.js]: If we decide to go all-in with good error handling (probably out-of-scope), maybe a ParseError?

File: packages/react-server-website/.reactserverrc [hunk 18]
```diff
@@ -1,19 +1,18 @@
 {
   "routes": "./routes.js",
   "host": "localhost",
-  "port": 3000,
-  "js-port": 3001,
+  "port": 3010,
```
Comment (doug-wade) [inline packages/react-server-website/.reactserverrc]: this should probably go in the manifest

File: docs/guides/react-server-cli.md (hunk 2 above)
Comment (doug-wade) [inline docs/guides/react-server-cli.md]: :facepalm: I totally thought I was in the README.md. 🐑

File: packages/react-server-cli/src/logProductionWarnings.js (hunk 16 above)
Comment (gigabo) [inline packages/react-server-cli/src/logProductionWarnings.js]: Actually probably better not to default here.  You may have specific configuration for `env.development`, but if your `NODE_ENV` is actually undefined it won't get picked up.

File: packages/react-server-website/.reactserverrc (hunk 18 above)
Comment (gigabo) [inline packages/react-server-website/.reactserverrc]: Which manifest?

File: packages/react-server-cli/src/parseCliArgs.js (hunk 17 above)
Comment (doug-wade) [inline packages/react-server-cli/src/parseCliArgs.js]: >  For ordinary errors where stack trace is actually desirable I don't see any need to subclass.

applies here too

**Comment by doug-wade** [comment]
I had a lot to say about this pr, but its super awesome and I like the new interface way better and I whole-heartedly endorse this plan.  Thanks for pulling this together @gigabo

File: packages/react-server-website/.reactserverrc (hunk 18 above)
Comment (doug-wade) [inline packages/react-server-website/.reactserverrc]: docs/testing-ports.md

Comment (gigabo) [inline packages/react-server-website/.reactserverrc]: Oh.  Yeah, if we ever add tests for the website we'll definitely need to update the manifest. ;)

**Comment by gigabo** [comment]
Gonna merge this because I want to build on it.

@aickin - If you get a chance please take a look!
